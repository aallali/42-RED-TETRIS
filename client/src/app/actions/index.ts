import {
	SET_PLAYER, SET_GAME,
	LOGOUT_PLAYER, UPDATE_SCORE,
	SET_PLAYER_ADMIN, ADD_ROW,
	PLAYER_LOST, RESET_STATES,
	BROADCAST_SCORE,
	UNSET_PLAYER_ADMIN
} from "../../reducers/player.reducer";
import { UPDATE_PLAYERS, CLEAR_OPPONENTS } from "../../reducers/opponent.reducer"
import { SET_ERROR, CLEAR_ERROR } from "../../reducers/error.reducer"

import {
	SHIFT_TETRO, MORE_TETROS,
	SET_GAME_MODE, SET_GAME_OVER,
	SET_GAME_STARTED, SET_IN_GAME,
	SET_WINNER, RESET_GAME,
	UPDATE_GAME_MODE,
	LEAVE_GAME,
	SET_GAME_TITLE
} from "../../reducers/game.reducer"

export {
	// Players actions
	SET_PLAYER, SET_GAME,
	LOGOUT_PLAYER,
	UPDATE_SCORE,
	SET_PLAYER_ADMIN,
	ADD_ROW, PLAYER_LOST,
	BROADCAST_SCORE,
	UNSET_PLAYER_ADMIN,
	RESET_STATES,

	// Opponents actions
	UPDATE_PLAYERS,
	CLEAR_OPPONENTS,
	// Error actions
	SET_ERROR, CLEAR_ERROR,

	// game actions
	SHIFT_TETRO, MORE_TETROS,
	SET_GAME_MODE, SET_GAME_OVER,
	SET_GAME_STARTED, SET_IN_GAME,
	SET_WINNER, RESET_GAME,
	UPDATE_GAME_MODE,
	LEAVE_GAME,
	SET_GAME_TITLE
}
