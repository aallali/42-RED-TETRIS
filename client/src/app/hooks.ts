import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';
import { useGameStatus } from "./TetrisHooks/useGameStatus"
import { useInterval } from "./TetrisHooks/useInterval"
import { usePlayer } from "./TetrisHooks/usePlayer"
import { useStage } from "./TetrisHooks/useStage"


import io from "socket.io-client";

export const socket = io("http://localhost:4242");
export { useGameStatus, usePlayer, useInterval, useStage }
// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
