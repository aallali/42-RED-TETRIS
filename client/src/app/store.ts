import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import playerReducer from "../reducers/player.reducer";
import stageReducer from "../reducers/stage.reducer"
export const store = configureStore({
  reducer: {
    player: playerReducer,
    stage: stageReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
