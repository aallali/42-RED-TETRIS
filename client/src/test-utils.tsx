// test-utils.jsx
import React, { ReactChildren } from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
// Import your own reducer
import playerReducer from "./reducers/player.reducer";
import opponentReducer from "./reducers/opponent.reducer"
import errorReducer from "./reducers/error.reducer"
import gameReducer from "./reducers/game.reducer"
import chatReducer from './reducers/chat.reducer';

function aallaliStore(preloadedState: any) {
	return configureStore({
		reducer: {
			player: playerReducer,
			opponents: opponentReducer,
			error: errorReducer,
			game: gameReducer,
			chat: chatReducer
		},
		preloadedState
	})
}
function render(
	ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
	{
		preloadedState,
		store = configureStore({
			reducer: {
				player: playerReducer,
				opponents: opponentReducer,
				error: errorReducer,
				game: gameReducer,
				chat: chatReducer
			},
			preloadedState
		}),
		...renderOptions
	}: any = {}
) {
	function Wrapper({ children }: { children: ReactChildren }) {
		return <Provider store={store}>{children}</Provider>

	}
	return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}


// re-export everything
export * from '@testing-library/react'
// override render method
export { render, aallaliStore }