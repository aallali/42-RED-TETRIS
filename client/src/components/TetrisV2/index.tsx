/* eslint-disable react-hooks/exhaustive-deps */

import React, { useEffect } from 'react';
import { createStage, isColliding } from '../../helpers/gameHelpers';

// Custom hooks
import { useGameStatus ,usePlayer, useStage, useInterval} from '../../app/hooks';

// Components
import Stage from './components/Stage/Stage';

// Styles

// Costum Constans
import { STAGE_HEIGHT, STAGE_WIDTH } from "../../helpers/tetrominos"
import { socket, useAppDispatch, useAppSelector } from '../../app/hooks';

// Actions
import { PLAYER_LOST, SET_GAME_OVER, SHIFT_TETRO, UPDATE_SCORE } from '../../app/actions';

// Selectors
import { isGameStarted as isGameStartedS, isGameOver, getTetros, } from "../../reducers/game.reducer"
import { getRows2Add } from '../../reducers/player.reducer';


const TetrisV2: React.FC = () => {
	const dispatch = useAppDispatch()

	const isGameStarted = useAppSelector(isGameStartedS)
	const gameOver = useAppSelector<boolean>(isGameOver)
	const rows2add = useAppSelector(getRows2Add)
	const tetros = useAppSelector(getTetros)

	const [dropTime, setDroptime] = React.useState<null | number>(null);
	const gameArea = React.useRef<HTMLDivElement>(null);

	const { player, updatePlayerPos, resetPlayer, playerRotate } = usePlayer();
	const { stage, setStage, rowsCleared } = useStage(player, resetPlayer, (tetros[0] || 0));
	const { score, setScore, rows, setRows, level, setLevel } = useGameStatus(rowsCleared);

	useEffect(() => {
		dispatch(UPDATE_SCORE({ level: level, rows: rows, score: score, stage: stage }))
	}, [rowsCleared])

	useEffect(() => {
		if (isGameStarted === true) {
			handleStartGame()
		}
	}, [isGameStarted])

	useEffect(() => {
		if (gameOver)
			setDroptime(null);
	}, [gameOver])
	useEffect(() => {
		if (isGameStarted && tetros.length <= 5) {
			socket.emit("TETROS_PLEASE")
		}
	}, [tetros.length, isGameStarted])

	useEffect(() => {
		if (isGameStarted) {
			let f = rows2add
			let tmp = new Array(10).fill(["X", "merged"]);
			for (let i = 0; i < STAGE_HEIGHT; i++)
				if (stage[i][0][0] === "X" && stage[i][0][1] === "merged") f--;

			if (f > 0) {
				while (f > 0) {
					updatePlayerPos({ x: 0, y: -1, collided: false });
					stage.push(tmp);
					stage.shift();
					f--;
				}
				dispatch(UPDATE_SCORE({ level: level, rows: rows, score: score, stage: stage }))
			}
		}
	}, [rows2add]);
	const verticalDrop = () => {
		let tmp = 0;
		for (let i = 0; i < STAGE_HEIGHT; i++) {
			if (isColliding(player, stage, { x: 0, y: i })) {
				tmp = i;
				break;
			}
		}
		for (let i = tmp; i > 0; i--) {
			if (!isColliding(player, stage, { x: 0, y: i })) {
				updatePlayerPos({ x: 0, y: i, collided: false });
				break;
			}
		}
		return
	};
	const movePlayer = (dir: number) => {
		if (!isColliding(player, stage, { x: dir, y: 0 })) {
			updatePlayerPos({ x: dir, y: 0, collided: false });
		}
	};

	const keyUp = ({ keyCode }: { keyCode: number }): void => {
		if (!gameOver) {
			// Change the droptime speed when user releases down arrow
			if (keyCode === 40) {
				setDroptime(1000 / level + 100);
			}
		}
	};

	const handleStartGame = (): void => {
		// Need to focus the window with the key events on start
		if (gameArea.current) gameArea.current.focus();
		// Reset everything
		setStage(createStage());
		setDroptime(1000);
		resetPlayer(0);
		setScore(0);
		setLevel(1);
		setRows(0);
	};

	const move = ({ keyCode, repeat }: { keyCode: number; repeat: boolean }): void => {
		if (!gameOver) {
			if (keyCode === 37) {
				movePlayer(-1);
			} else if (keyCode === 39) {
				movePlayer(1);
			} else if (keyCode === 40) {
				// Just call once
				if (repeat) return;
				setDroptime(30);
			} else if (keyCode === 38) {
				playerRotate(stage);
			} else if (keyCode === 32) {
				verticalDrop()
			}
		}
	};

	const drop = (): void => {
		// Increase level when player has cleared 10 rows
		if (rows > level + 1) {
			setLevel(prev => prev + 1);
			// Also increase speed
			setDroptime(1000 / level + 100);
		}

		if (!isColliding(player, stage, { x: 0, y: 1 })) {
			updatePlayerPos({ x: 0, y: 1, collided: false });
		} else {
			// Game over!
			if (player.pos.y < 1) {
				setDroptime(null);
				dispatch(PLAYER_LOST())
				// dispatch(SET_GAME_OVER())
			}
			dispatch(SHIFT_TETRO())
			updatePlayerPos({ x: 0, y: 0, collided: true });
			dispatch(UPDATE_SCORE({ level: level, rows: rows, score: score, stage: stage }))
		}
	};

	useInterval(() => {
		drop();
	}, dropTime);

	return (
		<div role='button' tabIndex={0} onKeyDown={move} onKeyUp={keyUp} ref={gameArea}>
			<Stage
				stage={stage}
				STAGE_WIDTH={STAGE_WIDTH}
				STAGE_HEIGHT={STAGE_HEIGHT}
				CELL_SIZE={23} />
		</div>
	);
};

export default TetrisV2;
