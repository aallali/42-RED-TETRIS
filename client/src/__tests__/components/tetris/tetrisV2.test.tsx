
import { Provider } from 'react-redux';
import { store } from '../../../app/store';
import TetrisV2 from '../../../components/TetrisV2';
import renderer from "react-test-renderer";
import * as reactRedux from "react-redux";
import { render, screen } from '../../../test-utils'
import { fireEvent, waitFor, waitForElementToBeRemoved } from '@testing-library/react';
import { rest } from "msw";
import { setupServer } from "msw/node";
import { mount } from 'enzyme';
// import { act } from 'react-test-renderer';
// import { act } from '@testing-library/react';
// import { act } from 'react-dom/test-utils';

const server = setupServer(
	rest.get("http://localhost:4242/rooms", (_, res, ctx) => {
		return res(ctx.status(200), ctx.json([{ title: "arena", size: 7, active_players: 5, mode: "multiplayer" }]))
	}),
	rest.get("http://localhost:4242/top", (_, res, ctx) => {
		return res(ctx.status(200), ctx.json(`allali,67\nbot2,30\nbotUsername3,27`))
	})
);

// jest.mock('axios')


describe("_ TetrisV2 Testing ...", () => {
	const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");
	beforeEach(() => {
		useDispatchMock.mockClear();
		useDispatchMock.mockReturnValue(jest.fn());
	});
	beforeAll(() => server.listen());

	afterEach(() => server.resetHandlers());

	afterAll(() => server.close());

	test("__TetrisV2 snapshot test", () => {
		const tree = renderer.create(<Provider store={store}>
			< TetrisV2 />
		</Provider>).toJSON();
		expect(tree).toMatchSnapshot();
	});

	test("__TetrisV2 invoke states value", () => {
		render(<TetrisV2 />, {
			preloadedState:
			{
				player: { nickname: "aallali", rows2add: 3 },
				game: { started: true, title: "arena", tetros: ['Z', 'T', 'L', 'O'] }
			}
		})
	});

	test("__TetrisV2 invoke states value", () => {
		render(<TetrisV2 />, {
			preloadedState:
			{
				player: { nickname: "aallali", rows2add: 3 },
				game: { started: false, title: "arena", tetros: ['Z', 'T', 'L', 'O'] }
			}
		})
	});

	test("__TetrisV2 invoke states value", () => {
		render(<TetrisV2 />, {
			preloadedState:
			{
				player: { nickname: "aallali", rows2add: 3 },
				game: { started: false, gameOver: true, title: "arena", tetros: ['Z', 'T', 'L', 'O'] }
			}
		})
	});
})
