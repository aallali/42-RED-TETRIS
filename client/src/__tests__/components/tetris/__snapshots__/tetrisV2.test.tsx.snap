// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_ TetrisV2 Testing ... __TetrisV2 snapshot test 1`] = `
<div
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    className="sc-gsDKAQ fTyBlK"
  >
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
    <div
      className="sc-bdvvtL hygDQU"
      color="244, 239, 201"
      type={0}
    />
  </div>
</div>
`;
